### ISSUE TITLE
Adding `from_map` like API for Dask Arrays

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#9049]
Recently @rjzamora did some work to add a `from_map` API for Dask DataFrames in PR ( https://github.com/dask/dask/pull/8911 ). It might be interesting to consider a similar API for Dask Array. The [approach taken in Dask-Image's `imread`]( https://github.com/dask/dask-image/blob/890933e35f4f03157a7c3074dfc3cf2de2ab8b2e/dask_image/imread/__init__.py#L76-L99 ) might be a starting point for thinking about this.

cc @dask/array

**Comment by ian-r-rose** [comment]
I feel like `from_array` does *most* of this, where instead of implementing a callable that produces a chunk, you implement a `__getitem__`. The major difference is that `from_map` is a bit more flexible in that it can take a list of iterables which will be broadcast against the callable, whereas `__getitem__` only takes block indices.  But I'm not sure that difference is insurmountable, since the arraylike could keep a dictionary mapping a block tuple to whatever you want (cf. `BlockwiseDepDict`).

Is there something that one could do with `from_map` that we couldn't do with `from_array`? Or would it mostly be a convenience function?

**Comment by jakirkham** [comment]
A common conversation that I and others have with users of Dask (Array or DataFrame) is, how do I get my data into Dask?

`from_array` is fine for the Zarr user, but there are still users out there that have a directory of `*` (TIFFs, CSVs, some custom format, etc.) that need some help getting things into Dask.

Usually we mention `delayed` and point them to [this blogpost]( https://blog.dask.org/2019/06/20/load-image-data ), which is good for getting them started. Though going through `delayed` can be a bit slower and take a bit more work for the user. It also doesn't benefit from HLGs (unless I'm missing something).

The code in Dask-Image linked above tried to improve on this by "using" `map_blocks` to coerce an array of filenames into the loaded array data, which gets pretty close to what an array user might want in terms of a reasonably performant loading path.

However that code is a bit fiddly and complex for the average user. Ideally they wouldn't need to know about those internals. Having some kind of `from_map_blocks` might get us a lot closer to what a user would want.

**Comment by ian-r-rose** [comment]
> It also doesn't benefit from HLGs (unless I'm missing something).

Right, though `from_array` does now use HLGs, cf. #7417. What I'm proposing is (I think) pretty similar to `from_map`, but it already exists today in `dask.array`. Something like the following (pseudocode since I've changed `_map_read_frame`):

```python
from dask.layers import AraySlicesDep

class Image:
    def __init__(filenames, chunks):
        self.slices = ArraySlicesDep(chunks)
        self.filenames = filenames

    def __getitem__(chunk):
        return _map_read_frame(self.filenames[chunk], array_slice=self.slices[chunk])

image = Image(filenames, chunks)
arr = da.from_array(image)
```

This is similar to providing a callable to `from_map`, is lazy, and works with HLGs. There's some trickery around reproducing the `block-info` data using `ArraySlicesDep`, but that's a problem that would need to be solved with `from_map` as well (we discussed whether `BlockwiseDep` should be an acceptable input to `from_map` in #8911, and decided to not do it for the time being).

**Comment by jakirkham** [comment]
Interesting thanks for composing that example 🙏

It would be nice if we baked this class into Dask and then took a handful of inputs to a function instead. Maybe we could handle some of the selection mapping internally for them as well.

**Comment by clbarnes** [comment]
> from_array is fine for the Zarr user, but there are still users out there that have a directory of * (TIFFs, CSVs, some custom format, etc.) that need some help getting things into Dask.

As it happens I was doing this yesterday. We are gradually migrating from directories of JPEGs to zarr/N5, so I wrote a zarr store which wraps around one of our common [JPEG stack structures](https://catmaid.readthedocs.io/en/stable/tile_sources.html#file-based-image-stack-with-zoom-level-directories). It's a bit of a hack but not too inconvenient; may be useful to someone who comes across this issue before Dask gets the functionality itself.

```python
import typing as tp
from pathlib import Path
import json

from PIL import Image
import zarr
import numpy as np

FMT = "{z}/{scale}/{y}_{x}{ext}"

zarr_meta = {
    "zarr_format": 2,
    # "shape": None,  # to fill
    # "chunks": None,  # to fill
    "dtype": "|u1",
    "compressor": None,
    # "fill_value": None,  # to fill
    "order": "C",
    "filters": None,
    "dimension_separator": ".",
}

class ImageStore4(zarr._storage.store.BaseStore):
    is_listable = False
    is_writable = False
    is_erasable = False

    def __init__(
        self,
        root: Path,
        shape: tp.Tuple[int, int, int],
        im_height_width: tp.Tuple[int, int],
        scale: int = 0,
        ext: str = ".jpg",
        fill_value: int = 0,
        attrs: tp.Optional[dict] = None,
    ):
        self.root = Path(root)
        meta = zarr_meta.copy()
        if len(shape) != 3:
            raise ValueError("Shape must have 3 items")
        if len(im_height_width) != 2:
            raise ValueError("Image height/width must have 2 items")
        meta["shape"] = shape
        meta["chunks"] = (1, *im_height_width)
        meta["fill_value"] = fill_value
        self.meta = json.dumps(meta).encode()
        self.scale = scale
        self.ext = ext
        self.attrs = json.dumps(attrs or dict()).encode()

    def __getitem__(self, key):
        if "/" in key:
            raise ValueError(f"{type(self)} does not support groups")
        if key == ".zattrs":
            return self.attrs
        if key == ".zarray":
            return self.meta
        z, y, x = key.split(".")
        path = self.root / FMT.format(z=z, scale=self.scale, y=y, x=x, ext=self.ext)
        try:
            img = Image.open(path)
        except FileNotFoundError:
            raise KeyError()
        return np.asarray(img).tobytes()

    def __delitem__(self, _):
        raise NotImplementedError("Store is not erasable")

    def __iter__(self):
        raise NotImplementedError("Store is not listable")

    def __len__(self):
        raise NotImplementedError("Store is not listable")

    def __setitem__(self):
        raise NotImplementedError("Store is not writable")
```

The NotImplementedError boilerplate presumably won't be necessary once zarr v3 comes about. Not having to deal with zarr metadata and being able to pass a function to do basically what `__getitem__` does here would be more convenient, though.
